### PR TITLE
[Core] Refactoring distance process

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
This refactoring separates the ApplyRayCastingProcess for determining in and out from calculate distance. Tests are added for the new porcess.

### CODE REVIEW

**Comment by philbucher** [comment]
travis:
~~~
In file included from /home/travis/build/KratosMultiphysics/Kratos/cmake_build/kratos/cotire/KratosCore_CXX_unity_0_97.cxx:49:0:
/home/travis/build/KratosMultiphysics/Kratos/kratos/processes/calculate_distance_to_skin_process.cpp: In member function ‘virtual void Kratos::CalculateDistanceToSkinProcess<TDim>::CalculateRayDistances()’:
/home/travis/build/KratosMultiphysics/Kratos/kratos/processes/calculate_distance_to_skin_process.cpp:191:3: error: ‘ApplyRayCastingProcess’ was not declared in this scope
   ApplyRayCastingProcess<TDim> ray_casting_process(CalculateDiscontinuousDistanceToSkinProcess<TDim>::mFindIntersectedObjectsProcess, mExtraRaysEpsilon);
   ^~~~~~~~~~~~~~~~~~~~~~
In file included from /home/travis/build/KratosMultiphysics/Kratos/cmake_build/kratos/cotire/KratosCore_CXX_unity_0_97.cxx:49:0:
/home/travis/build/KratosMultiphysics/Kratos/kratos/processes/calculate_distance_to_skin_process.cpp:192:3: error: ‘ray_casting_process’ was not declared in this scope
   ray_casting_process.Execute();
   ^~~~~~~~~~~~~~~~~~~
~~~

File: kratos/processes/apply_ray_casting_process.h [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    const double mExtraRaysEpsilon = 1.0e-8;
+    FindIntersectedGeometricalObjectsProcess* mpFindIntersectedObjectsProcess;
+    bool mIsSearchStructureAllocated;
+    double mCharactresticLength;
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.h]:
```suggestion
    double mCharacteristicLength;
```
(typo)

File: kratos/processes/apply_ray_casting_process.h [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+     * @param pRayPoint2 ray end point
+     * @param pIntersectionPoint obtained intersecting point
+     * @return int integer containing the result of the intersection (1 if intersects)
+     */
+    int ComputeRayIntersection(
+        Element::GeometryType& rGeometry,
+        const double* pRayPoint1,
+        const double* pRayPoint2,
+        double* pIntersectionPoint);
+
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.h:216]: I'd take the chance to add the Doxygen to all these methods below.

File: kratos/processes/apply_ray_casting_process.cpp [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+            mpFindIntersectedObjectsProcess->Initialize();
+
+        CalculateCharacteristicLength();
+
+		ModelPart& ModelPart1 = mpFindIntersectedObjectsProcess->GetModelPart1();
+
+		#pragma omp parallel for
+		for(int k = 0 ; k < static_cast<int>(ModelPart1.NumberOfNodes()); ++k) {
+			auto it_node = ModelPart1.NodesBegin() + k;
+			double &node_distance = it_node->GetSolutionStepValue(DISTANCE);
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
			double &r_node_distance = it_node->GetSolutionStepValue(DISTANCE);
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+			const double ray_distance = this->DistancePositionInSpace(*it_node);
+			if (ray_distance * node_distance < 0.0) {
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
			if (ray_distance * r_node_distance < 0.0) {
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+				node_distance = -node_distance;
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
				r_node_distance = -r_node_distance;
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+	}
+
+	template<std::size_t TDim>
+	double ApplyRayCastingProcess<TDim>::DistancePositionInSpace(const Node<3> &rNode)
+	{
+        const double epsilon = 1e-12;
+		array_1d<double,TDim> distances;
+		unsigned int n_ray_pos(0), n_ray_neg(0);
+        IntersectionsContainerType intersections;
+		const array_1d<double,3> &r_coords = rNode.Coordinates();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
		const auto &r_coords = rNode.Coordinates();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+
+		// Loop the x,y and z (3D) ray directions
+        for (unsigned int i_direction = 0; i_direction < TDim; i_direction++){
+			// Initialize the current direction distance
+			distances[i_direction] = 1.0;
+
+            // Creating the ray
+            double ray[3] = {r_coords[0], r_coords[1], r_coords[2]};
+			OctreeType* pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
			auto pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+            pOctree->NormalizeCoordinates(ray);
+            ray[i_direction] = 0; // Starting from the lower extreme
+
+            this->GetRayIntersections(ray, i_direction, intersections);
+
+            int ray_color = 1;
+            std::vector<std::pair<double, Element::GeometryType*> >::iterator i_intersection = intersections.begin();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
            auto i_intersection = intersections.begin();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+		const double RayPerturbation,
+		const array_1d<double,3> &rCoords,
+        array_1d<double,TDim> &rDistances)
+	{
+		// Set the extra ray origins
+		std::vector<array_1d<double,3>> extra_ray_origs;
+		this->GetExtraRayOrigins(RayPerturbation, rCoords, extra_ray_origs);
+
+		// Get the pointer to the base Octree binary
+		OctreeType* p_octree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
		auto p_octree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+
+		// Loop the extra rays to compute its color
+		unsigned int n_ray_pos = 0; // Positive rays counter
+		unsigned int n_ray_neg = 0; // Negative rays counter
+		IntersectionsContainerType intersections; // Ray intersections container initialization
+		for (unsigned int i_direction = 0; i_direction < TDim; ++i_direction) {
+			for (unsigned int i_ray = 0; i_ray < extra_ray_origs.size(); ++i_ray) {
+				// Creating the ray
+				array_1d<double,3> aux_ray = extra_ray_origs[i_ray];
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
				const auto aux_ray = extra_ray_origs[i_ray];
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+				double ray[3] = {aux_ray[0], aux_ray[1], aux_ray[2]};				
+				p_octree->NormalizeCoordinates(ray);
+				ray[i_direction] = 0; // Starting from the lower extreme
+				this->CorrectExtraRayOrigin(ray); // Avoid extra ray normalized coordinates to be larger than 1 or 0
+				this->GetRayIntersections(ray, i_direction, intersections);
+
+				// Compute the extra rays intersections
+				int ray_color = 1;
+				std::vector<std::pair<double, Element::GeometryType*> >::iterator i_intersection = intersections.begin();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
				auto i_intersection = intersections.begin();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+				while (i_intersection != intersections.end()) {
+					double int_d = extra_ray_origs[i_ray][i_direction] - i_intersection->first; // Octree ray intersection distance
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
					const double int_d = extra_ray_origs[i_ray][i_direction] - i_intersection->first; // Octree ray intersection distance
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+        // Ray is of dimension (3) normalized in (0,1)^3 space
+        // Direction can be 0,1,2 which are x,y and z respectively
+
+        const double epsilon = 1.00e-12;
+
+        // First clearing the intersections points vector
+        rIntersections.clear();
+
+		// Get the octree from the parent discontinuous distance process
+        OctreeType* pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
        auto pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+
+		// Compute the normalized ray key
+        OctreeType::key_type ray_key[3] = {pOctree->CalcKeyNormalized(ray[0]), pOctree->CalcKeyNormalized(ray[1]), pOctree->CalcKeyNormalized(ray[2])};
+
+        // Getting the entrance cell from lower extreme
+        OctreeType::key_type cell_key[3];
+        OctreeType::cell_type* cell = pOctree->pGetCell(ray_key);
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
        auto cell = pOctree->pGetCell(ray_key);
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 15]
```diff
@@ -0,0 +1,485 @@
...
+                cell = NULL;
+			}
+        }
+
+        // Eliminate the repeated intersecting objects
+        if (!rIntersections.empty()) {
+            // Sort
+            std::sort(rIntersections.begin(), rIntersections.end());
+            // Unique
+            std::vector<std::pair<double, Element::GeometryType*> >::iterator i_begin = rIntersections.begin();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
            auto i_begin = rIntersections.begin();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+            std::vector<std::pair<double, Element::GeometryType*> >::iterator i_intersection = rIntersections.begin();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
            auto i_intersection = rIntersections.begin();
```

File: kratos/processes/apply_ray_casting_process.cpp [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+		const double* ray,
+		OctreeType::key_type* ray_key,
+		const unsigned int direction,
+		std::vector<std::pair<double, Element::GeometryType*> > &rIntersections)
+	{
+		//This function passes the ray through the cell and gives the hit point to all objects in its way
+		//ray is of dimension (3) normalized in (0,1)^3 space
+		// direction can be 0,1,2 which are x,y and z respectively
+		typedef OctreeType::cell_type::object_container_type object_container_type;
+		object_container_type* objects = (cell->pGetObjects());
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
		auto objects = (cell->pGetObjects());
```
(if this works I think we can remove the `typedef` above)

File: kratos/processes/apply_ray_casting_process.cpp [hunk 18]
```diff
@@ -0,0 +1,485 @@
...
+		if (objects->empty()){
+			return 0;
+		}
+
+		// Calculating the two extreme of the ray segment inside the cell
+		double ray_point1[3] = {ray[0], ray[1], ray[2]};
+		double ray_point2[3] = {ray[0], ray[1], ray[2]};
+		double normalized_coordinate;
+
+		OctreeType* pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]:
```suggestion
		auto pOctree = mpFindIntersectedObjectsProcess->GetOctreePointer();
```

File: kratos/processes/apply_ray_casting_process.cpp (hunk 17 above)
Comment (rubenzorrilla) [reply kratos/processes/apply_ray_casting_process.cpp]: Actually the `typedef` is needed in a loop below if we don't change it to an auto loop.

File: kratos/processes/apply_ray_casting_process.cpp [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+		array_1d<double,3> ray_pt_1;
+		array_1d<double,3> ray_pt_2;
+		for (unsigned int i = 0; i < 3; ++i){
+			ray_pt_1[i] = pRayPoint1[i];
+			ray_pt_2[i] = pRayPoint2[i];
+		}
+
+		// Call the line - line intersection util
+		array_1d<double,3> int_pt = ZeroVector(3);
+		const double tolerance = 1.0e-6*rGeometry.Length();
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp:388]: This is just to note that this tolerance should be changed. This has been tried in commit 98c1b250efc1b5ce713b00a23ee5d0fe6904ef04 .

File: kratos/processes/apply_ray_casting_process.cpp [hunk 20]
```diff
@@ -0,0 +1,485 @@
...
+		array_1d<double,3> ray_pt_1;
+		array_1d<double,3> ray_pt_2;
+		for (unsigned int i = 0; i < 3; ++i){
+			ray_pt_1[i] = pRayPoint1[i];
+			ray_pt_2[i] = pRayPoint2[i];
+		}
+
+		// Call the line - triangle intersection util
+		array_1d<double,3> int_pt = ZeroVector(3);
+		const double tolerance = 1.0e-6*std::sqrt(rGeometry.Length());
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp:421]: Same comment.

File: kratos/processes/apply_ray_casting_process.cpp [hunk 21]
```diff
@@ -0,0 +1,485 @@
...
+                min_point[i]  =  (min_point[i]  >  node[i] ) ?  node[i] : min_point[i];
+                max_point[i] =  (max_point[i] <  node[i] ) ?  node[i] : max_point[i];
+            }            
+        }
+
+        double distance = norm_2(max_point - min_point);
+        if (distance > std::numeric_limits<double>::epsilon())
+            mCharactresticLength = distance;
+
+        KRATOS_WATCH(mCharactresticLength);
```
Comment (rubenzorrilla) [inline kratos/processes/apply_ray_casting_process.cpp]: I think that you forget to remove this one.

**Review by rubenzorrilla — COMMENTED** [review]
Since we are adding a new file, I think it have sense to take the chance to do the clean up modifications suggested as well as to properly format the file.

File: kratos/processes/apply_ray_casting_process.cpp (hunk 19 above)
Comment (pooyan-dadvand) [reply kratos/processes/apply_ray_casting_process.cpp:388]: But I concluded that we change the tolerance in separate PR to make the refactoring without behavior change

**Comment by pooyan-dadvand** [comment]
> Since we are adding a new file, I think it have sense to take the chance to do the clean up modifications suggested as well as to properly format the file.

I have already passed it to the clang format. Where is not correct?

**Comment by rubenzorrilla** [comment]
> > Since we are adding a new file, I think it have sense to take the chance to do the clean up modifications suggested as well as to properly format the file.
> 
> I have already passed it to the clang format. Where is not correct?

Yesterday GitHub file reader showed lots of issues with the indentation.

File: kratos/processes/apply_ray_casting_process.cpp (hunk 19 above)
Comment (rubenzorrilla) [reply kratos/processes/apply_ray_casting_process.cpp:388]: Yes, we agreed on that. This was just a reminder :wink:

**Comment by pooyan-dadvand** [comment]
So if you are ok with changes would you pleaes approve it?

**Review by rubenzorrilla — APPROVED** [review]
Please go ahead

**Pull request merged**